src/dynmem.hpp

```diff
@@ -15,110 +15,191 @@
 #endif
 #endif
 
-class MemoryManager {
+#if !defined(CUFFT) && defined(CUDA_TRAD)
+#error "CUFFT needs to be enabled if CUDA_TRAD is used"
+#endif
+
+#ifdef CUDA_TRAD
+enum MemLoc{
+    CURR_HOST,
+    CURR_DEV
+};
+#endif /* CUDA_TRAD */
+
+template <typename T>
+struct MemNode {
+  T *ptr_h;
+#ifdef CUFFT
+  T *ptr_d;
+#ifdef CUDA_TRAD
+  mutable MemLoc currLoc;
+#endif /* CUDA_TRAD */
+#endif /* CUFFT */
+};
+
+template <typename T> class MemoryManager_ {
     std::mutex mutex;
-    std::map<size_t, std::stack<void*> > map;
+    std::map<size_t, std::stack<T> > map;
 
 public:
-    void *get(size_t size) {
+    bool get(T *node, size_t size) {
         std::lock_guard<std::mutex> guard(mutex);
         auto &stack = map[size];
-        void *ptr = nullptr;
         if (!stack.empty()) {
-            ptr = stack.top();
+            *node = stack.top();
             stack.pop();
+            return true;
         }
-        return ptr;
+        return false;
     }
-    void put(void *ptr, size_t size) {
+    void put(T *node, size_t size) {
         std::lock_guard<std::mutex> guard(mutex);
-        map[size].push(ptr);
+        map[size].push(*node);
     }
 };
 
+typedef MemoryManager_<struct MemNode<void>> MemoryManager;
+
 template <typename T> class DynMem_ {
   private:
-    T *ptr_h = nullptr;
-#ifdef CUFFT
-    T *ptr_d = nullptr;
+    struct MemNode<T> mem;
     static MemoryManager mmng;
-#endif
   public:
     typedef T value_type;
     const size_t num_elem;
 
     DynMem_(size_t num_elem) : num_elem(num_elem)
     {
+        if (!mmng.get(reinterpret_cast<struct MemNode<void>*>(&mem), num_elem))
 #ifdef CUFFT
-        ptr_h = reinterpret_cast<T*>(mmng.get(num_elem));
-        if (!ptr_h)
-            CudaSafeCall(cudaHostAlloc(reinterpret_cast<void **>(&ptr_h), num_elem * sizeof(T), cudaHostAllocMapped));
+#ifdef CUDA_TRAD
+        {
+            mem.currLoc = CURR_HOST;
+            mem.ptr_h = new T[num_elem];
+            CudaSafeCall(cudaMalloc(&mem.ptr_d, num_elem * sizeof(T)));
+        }
+#else /* CUDA_TRAD */
+        {
+            CudaSafeCall(cudaHostAlloc(reinterpret_cast<void **>(&mem.ptr_h), num_elem * sizeof(T), cudaHostAllocMapped));
+        }
 
-        CudaSafeCall(cudaHostGetDevicePointer(reinterpret_cast<void **>(&ptr_d), reinterpret_cast<void *>(ptr_h), 0));
-#else
-        ptr_h = new T[num_elem];
-#endif
+        CudaSafeCall(cudaHostGetDevicePointer(reinterpret_cast<void **>(&mem.ptr_d), reinterpret_cast<void *>(mem.ptr_h), 0));
+#endif /* CUDA_TRAD */
+#else /* CUFFT */
+        {
+            mem.ptr_h = new T[num_elem];
+        }
+#endif /* CUFFT */
     }
     DynMem_(const DynMem_ &other) : DynMem_(other.num_elem)
     {
-        memcpy(ptr_h, other.ptr_h, num_elem * sizeof(T));
+        memcpy(mem.ptr_h, other.mem.ptr_h, num_elem * sizeof(T));
+#ifdef CUDA_TRAD
```

Comment:
How often is that called? If often, we might want to speed it up by copying only CPU or GPU memory based on the current owner.